### PR TITLE
fix(worker): fail on garbage incomplete leases

### PR DESCRIPTION
**Pull request by bencvdb**
Addresses https://fivetran.atlassian.net/browse/RD-101800 and is required for fivetran/engineering#97241.

### CODE REVIEW

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/leases/impl/KinesisClientLease.java [hunk 1]
```diff
@@ -71,6 +71,10 @@ public KinesisClientLease(KinesisClientLease other) {
         setParentShardIds(casted.parentShardIds);
     }
 
+    public boolean isCompete() {
```
Comment (fivetran-alvinchang) [inline src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/leases/impl/KinesisClientLease.java]: nit: sp(?) on isComplete?

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java [hunk 2]
```diff
@@ -552,6 +557,42 @@ private boolean causedByStreamRecordProcessingError(Throwable t) {
         return causedByStreamRecordProcessingError(t.getCause());
     }
 
+    private void handleTrimmedDataAccessException(Exception e) {
+        Optional<TrimmedDataAccessException> maybeTrimmedException = getTrimmedDataAccessException(e);
+        if (maybeTrimmedException.isPresent()) throw maybeTrimmedException.get();
+    }
+
+    private Optional<TrimmedDataAccessException> getTrimmedDataAccessException(Throwable t) {
```
Comment (fivetran-alvinchang) [inline src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: nit: It looks like we use this recursive way to get exception that we care about here, either `TrimmedDataAccessException` or `MissingIncompleteLeasesException`, but it looks like they are more or less doing the same thing, just checking for a different exception class. Could we parameterize the exception class and just use one method here?

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShardSyncer.java [hunk 3]
```diff
@@ -694,7 +702,7 @@ private static synchronized void cleanupLeasesOfFinishedShards(Collection<Kinesi
         Set<String> shardIdsOfClosedShards = new HashSet<>();
         List<KinesisClientLease> leasesOfClosedShards = new ArrayList<>();
         for (KinesisClientLease lease : currentLeases) {
-            if (lease.getCheckpoint().equals(ExtendedSequenceNumber.SHARD_END)) {
+            if (lease.isCompete()) {
```
Comment (fivetran-alvinchang) [inline src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShardSyncer.java]: nit: sp(?)

**Review by fivetran-alvinchang — APPROVED** [review]
LGTM besides some minor comments

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/leases/impl/KinesisClientLease.java (hunk 1 above)
Comment (bencvdb) [reply src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/leases/impl/KinesisClientLease.java]: Good catch! I'll fix this.

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java (hunk 2 above)
Comment (bencvdb) [reply src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: Nice call. I will fix this.

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java [hunk 4]
```diff
@@ -484,6 +486,10 @@ public void run() {
         try {
             initialize();
             LOG.info("Initialization complete. Starting worker loop.");
+        } catch (MissingIncompleteLeasesException e1){
+            LOG.error("Unable to initialize because some incomplete leases were missing. Shutting down.", e1);
+            shutdown();
+            throw e1;
```
Comment (bencvdb) [inline src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:492]: This is needed to make sure the exception can be caught by the connector.

**Review by fivetran-alvinchang — APPROVED** [review]

**Pull request merged**